Simplify code with decltype(auto)

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30924 Reimplement c10::apply / c10::invoke with c++14
* **#30922 Simplify code with decltype(auto)**

New c++14 feature we can use now

Differential Revision: [D18869644](https://our.internmc.facebook.com/intern/diff/D18869644/)

smessmer committed:
Simplify code with decltype(auto)

New c++14 feature we can use now

Differential Revision: [D18869644](https://our.internmc.facebook.com/intern/diff/D18869644/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify code with decltype(auto)"


New c++14 feature we can use now

Differential Revision: [D18869644](https://our.internmc.facebook.com/intern/diff/D18869644/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify code with decltype(auto)"


New c++14 feature we can use now

Differential Revision: [D18869644](https://our.internmc.facebook.com/intern/diff/D18869644/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify code with decltype(auto)"


New c++14 feature we can use now

Differential Revision: [D18869644](https://our.internmc.facebook.com/intern/diff/D18869644/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify code with decltype(auto)"


New c++14 feature we can use now

Differential Revision: [D18869644](https://our.internmc.facebook.com/intern/diff/D18869644/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify code with decltype(auto)"


New c++14 feature we can use now

Differential Revision: [D18869644](https://our.internmc.facebook.com/intern/diff/D18869644/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify code with decltype(auto)"


New c++14 feature we can use now

Differential Revision: [D18869644](https://our.internmc.facebook.com/intern/diff/D18869644/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify code with decltype(auto)"


New c++14 feature we can use now

Differential Revision: [D18869644](https://our.internmc.facebook.com/intern/diff/D18869644/)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 0377e32:
* **1/7** broken upstream at merge base 409151e (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=409151e1bb90c66546901d372f59930f72242ee9&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
    * You may want to rebase on the `viable/strict` branch (see its [recency history](https://dr.pytorch.org/master-viable-commits.html)):
        * If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
          ```
          git fetch viable/strict
          git rebase --onto viable/strict $(git merge-base origin/master HEAD)
          ```
        * If your commit is older than `viable/strict`:
          ```
          git fetch viable/strict
          git rebase viable/strict
          ```
* **6/7** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=0377e32ed2dac6370e158d9c8bc9b3d02f9ef192).

### 6 new failures recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3962449) caffe2_onnx_py2_gcc5_ubuntu16_04_build (1/6)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74969236))
<details>
<summary>
<code>Dec 16 10:18:25 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected string-literal before ')' token</code>
</summary>

```
Dec 16 10:18:25                                                                  ^ 
Dec 16 10:18:25 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:105:65: error: expected string-literal before ')' token 
Dec 16 10:18:25 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp: In member function 'virtual void {anonymous}::test_if_constexpr::if_constexpr_returnTypeCanDiffer_withoutIdentityArg_Test::TestBody()': 
Dec 16 10:18:25 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:113:71: error: expected ',' before ')' token 
Dec 16 10:18:25      static_assert(std::is_same<std::string, decltype(a_string)>::value); 
Dec 16 10:18:25                                                                        ^ 
Dec 16 10:18:25 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:113:71: error: expected string-literal before ')' token 
Dec 16 10:18:25 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected ',' before ')' token 
Dec 16 10:18:25      static_assert(std::is_same<int64_t, decltype(an_int)>::value); 
Dec 16 10:18:25                                                                  ^ 
Dec 16 10:18:25 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected string-literal before ')' token 
Dec 16 10:18:25 make[2]: *** [c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o] Error 1 
Dec 16 10:18:25 c10/test/CMakeFiles/c10_C++17_test.dir/build.make:62: recipe for target 'c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o' failed 
Dec 16 10:18:25 CMakeFiles/Makefile2:2595: recipe for target 'c10/test/CMakeFiles/c10_C++17_test.dir/all' failed 
Dec 16 10:18:25 make[1]: *** [c10/test/CMakeFiles/c10_C++17_test.dir/all] Error 2 
Dec 16 10:18:25 make[1]: *** Waiting for unfinished jobs.... 
Dec 16 10:18:25 [ 41%] Generating src/x86_64-fma/blas/sdotxf.py.o 
Dec 16 10:18:26 [ 41%] Generating src/x86_64-fma/blas/shdotxf.py.o 
Dec 16 10:18:29 Scanning dependencies of target nnpack 
Dec 16 10:18:29 [ 41%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/init.c.o 
Dec 16 10:18:29 [ 41%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/convolution-inference.c.o 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3962464) pytorch_linux_xenial_py3_6_gcc5_4_build (2/6)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74969302))
<details>
<summary>
<code>Dec 16 10:18:37 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected string-literal before ')' token</code>
</summary>

```
Dec 16 10:18:37                                                                  ^ 
Dec 16 10:18:37 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:105:65: error: expected string-literal before ')' token 
Dec 16 10:18:37 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp: In member function 'virtual void {anonymous}::test_if_constexpr::if_constexpr_returnTypeCanDiffer_withoutIdentityArg_Test::TestBody()': 
Dec 16 10:18:37 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:113:71: error: expected ',' before ')' token 
Dec 16 10:18:37      static_assert(std::is_same<std::string, decltype(a_string)>::value); 
Dec 16 10:18:37                                                                        ^ 
Dec 16 10:18:37 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:113:71: error: expected string-literal before ')' token 
Dec 16 10:18:37 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected ',' before ')' token 
Dec 16 10:18:37      static_assert(std::is_same<int64_t, decltype(an_int)>::value); 
Dec 16 10:18:37                                                                  ^ 
Dec 16 10:18:37 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected string-literal before ')' token 
Dec 16 10:18:37 c10/test/CMakeFiles/c10_C++17_test.dir/build.make:62: recipe for target 'c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o' failed 
Dec 16 10:18:37 make[2]: *** [c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o] Error 1 
Dec 16 10:18:37 CMakeFiles/Makefile2:2396: recipe for target 'c10/test/CMakeFiles/c10_C++17_test.dir/all' failed 
Dec 16 10:18:37 make[1]: *** [c10/test/CMakeFiles/c10_C++17_test.dir/all] Error 2 
Dec 16 10:18:37 make[1]: *** Waiting for unfinished jobs.... 
Dec 16 10:18:39 [ 42%] Linking CXX executable ../../bin/c10_typeid_test 
Dec 16 10:18:39 [ 42%] Built target c10_typeid_test 
Dec 16 10:18:44 [ 42%] Generating src/x86_64-fma/2d-winograd-8x8-3x3.py.o 
Dec 16 10:18:45 [ 42%] Generating src/x86_64-fma/blas/s8gemm.py.o 
Dec 16 10:18:46 [ 42%] Generating src/x86_64-fma/blas/c8gemm.py.o 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3962324) binary_linux_libtorch_2_7m_cpu_gcc5_4_cxx11-abi_shared-with-deps_build (3/6)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74970890))
<details>
<summary>
<code>Dec 16 10:19:08 FAILED: c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o</code>
</summary>

```
Dec 16 10:19:07 [752/2999] Building CXX object c10/test/CMakeFiles/c10_tempfile_test.dir/util/tempfile_test.cpp.o 
Dec 16 10:19:07 [753/2999] Linking CXX executable bin/c10_tempfile_test 
Dec 16 10:19:07 [754/2999] Building CXX object c10/test/CMakeFiles/c10_ConstexprCrc_test.dir/util/ConstexprCrc_test.cpp.o 
Dec 16 10:19:07 [755/2999] Building CXX object c10/test/CMakeFiles/c10_TensorTypeSet_test.dir/core/TensorTypeSet_test.cpp.o 
Dec 16 10:19:07 [756/2999] Linking CXX executable bin/c10_ConstexprCrc_test 
Dec 16 10:19:07 [757/2999] Linking CXX executable bin/c10_TensorTypeSet_test 
Dec 16 10:19:07 [758/2999] Building CXX object c10/test/CMakeFiles/c10_string_view_test.dir/util/string_view_test.cpp.o 
Dec 16 10:19:07 [759/2999] Linking CXX executable bin/c10_string_view_test 
Dec 16 10:19:08 [760/2999] Building CXX object c10/test/CMakeFiles/c10_TypeTraits_test.dir/util/TypeTraits_test.cpp.o 
Dec 16 10:19:08 [761/2999] Building CXX object c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o 
Dec 16 10:19:08 FAILED: c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o  
ror=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -O3  -fPIE   -DCAFFE2_USE_GLOO -DHAVE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -std=gnu++14 -MD -MT c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o -MF c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o.d -o c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o -c ../c10/test/util/C++17_test.cpp 
Dec 16 10:19:08 ../c10/test/util/C++17_test.cpp: In member function ‘virtual void {anonymous}::test_if_constexpr::if_constexpr_returnTypeCanDiffer_withIdentityArg_Test::TestBody()’: 
Dec 16 10:19:08 ../c10/test/util/C++17_test.cpp:99:71: error: expected ‘,’ before ‘)’ token 
Dec 16 10:19:08      static_assert(std::is_same<std::string, decltype(a_string)>::value); 
Dec 16 10:19:08                                                                        ^ 
Dec 16 10:19:08 ../c10/test/util/C++17_test.cpp:99:71: error: expected string-literal before ‘)’ token 
Dec 16 10:19:08 ../c10/test/util/C++17_test.cpp:105:65: error: expected ‘,’ before ‘)’ token 
Dec 16 10:19:08      static_assert(std::is_same<int64_t, decltype(an_int)>::value); 
Dec 16 10:19:08                                                                  ^ 
Dec 16 10:19:08 ../c10/test/util/C++17_test.cpp:105:65: error: expected string-literal before ‘)’ token 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3962461) pytorch_linux_xenial_cuda9_cudnn7_py3_build (4/6)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74974181))
<details>
<summary>
<code>Dec 16 10:21:23 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected string-literal before ')' token</code>
</summary>

```
Dec 16 10:21:23                                                                  ^ 
Dec 16 10:21:23 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:105:65: error: expected string-literal before ')' token 
Dec 16 10:21:23 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp: In member function 'virtual void {anonymous}::test_if_constexpr::if_constexpr_returnTypeCanDiffer_withoutIdentityArg_Test::TestBody()': 
Dec 16 10:21:23 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:113:71: error: expected ',' before ')' token 
Dec 16 10:21:23      static_assert(std::is_same<std::string, decltype(a_string)>::value); 
Dec 16 10:21:23                                                                        ^ 
Dec 16 10:21:23 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:113:71: error: expected string-literal before ')' token 
Dec 16 10:21:23 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected ',' before ')' token 
Dec 16 10:21:23      static_assert(std::is_same<int64_t, decltype(an_int)>::value); 
Dec 16 10:21:23                                                                  ^ 
Dec 16 10:21:23 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected string-literal before ')' token 
Dec 16 10:21:23 make[2]: *** [c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o] Error 1 
Dec 16 10:21:23 c10/test/CMakeFiles/c10_C++17_test.dir/build.make:62: recipe for target 'c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o' failed 
Dec 16 10:21:23 CMakeFiles/Makefile2:2736: recipe for target 'c10/test/CMakeFiles/c10_C++17_test.dir/all' failed 
Dec 16 10:21:23 make[1]: *** [c10/test/CMakeFiles/c10_C++17_test.dir/all] Error 2 
Dec 16 10:21:23 make[1]: *** Waiting for unfinished jobs.... 
Dec 16 10:21:24 Scanning dependencies of target nnpack 
Dec 16 10:21:24 [ 35%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/init.c.o 
Dec 16 10:21:24 [ 35%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/convolution-inference.c.o 
Dec 16 10:21:24 [ 35%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/fully-connected-inference.c.o 
Dec 16 10:21:24 [ 35%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/pooling-output.c.o 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3962460) pytorch_windows_build (5/6)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74974448))
<details>
<summary>
<code>FAILED: c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.obj</code>
</summary>

```
 
.281\bin\Hostx64\x64\link.exe  c10\test\CMakeFiles\c10_TypeTraits_test.dir\util\TypeTraits_test.cpp.obj  /out:bin\c10_TypeTraits_test.exe /implib:LIB\c10_TypeTraits_test.lib /pdb:bin\c10_TypeTraits_test.pdb /version:0.0  /machine:x64 /ignore:4049 /ignore:4217 /INCREMENTAL:NO /subsystem:console  LIB\c10.lib  LIB\gmock.lib  LIB\gtest.lib  LIB\gtest_main.lib  LIB\gtest.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ." 
Microsoft (R) Incremental Linker Version 14.23.28107.0 
Copyright (C) Microsoft Corporation.  All rights reserved. 
 
.exe  c10\test\CMakeFiles\c10_Metaprogramming_test.dir\util\Metaprogramming_test.cpp.obj  /out:bin\c10_Metaprogramming_test.exe /implib:LIB\c10_Metaprogramming_test.lib /pdb:bin\c10_Metaprogramming_test.pdb /version:0.0  /machine:x64 /ignore:4049 /ignore:4217 /INCREMENTAL:NO /subsystem:console  LIB\c10.lib  LIB\gmock.lib  LIB\gtest.lib  LIB\gtest_main.lib  LIB\gtest.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ." 
Microsoft (R) Incremental Linker Version 14.23.28107.0 
Copyright (C) Microsoft Corporation.  All rights reserved. 
 
etest\googlemock\include -I..\third_party\googletest\googlemock -I..\third_party\googletest\googletest\include -I..\third_party\googletest\googletest /DWIN32 /D_WINDOWS /GR  /w /EHa /MP /bigobj -openmp:experimental /MD /O2 /Ob2  /w /EHa /MP /bigobj   -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -std:c++14 /showIncludes /Foc10\test\CMakeFiles\c10_C++17_test.dir\util\C++17_test.cpp.obj /Fdc10\test\CMakeFiles\c10_C++17_test.dir\ /FS -c ..\c10\test\util\C++17_test.cpp 
FAILED: c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.obj  
etest\googlemock\include -I..\third_party\googletest\googlemock -I..\third_party\googletest\googletest\include -I..\third_party\googletest\googletest /DWIN32 /D_WINDOWS /GR  /w /EHa /MP /bigobj -openmp:experimental /MD /O2 /Ob2  /w /EHa /MP /bigobj   -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -std:c++14 /showIncludes /Foc10\test\CMakeFiles\c10_C++17_test.dir\util\C++17_test.cpp.obj /Fdc10\test\CMakeFiles\c10_C++17_test.dir\ /FS -c ..\c10\test\util\C++17_test.cpp 
..\c10\test\util\C++17_test.cpp(99): error C2429: language feature 'terse static assert' requires compiler flag '/std:c++17'
..\c10\test\util\C++17_test.cpp(105): error C2429: language feature 'terse static assert' requires compiler flag '/std:c++17'
..\c10\test\util\C++17_test.cpp(113): error C2429: language feature 'terse static assert' requires compiler flag '/std:c++17'
..\c10\test\util\C++17_test.cpp(119): error C2429: language feature 'terse static assert' requires compiler flag '/std:c++17'
Microsoft (R) C/C++ Optimizing Compiler Version 19.23.28107 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

VC\Tools\MSVC\1423~1.281\bin\Hostx64\x64\link.exe  c10\test\CMakeFiles\c10_either_test.dir\util\either_test.cpp.obj  /out:bin\c10_either_test.exe /implib:LIB\c10_either_test.lib /pdb:bin\c10_either_test.pdb /version:0.0  /machine:x64 /ignore:4049 /ignore:4217 /INCREMENTAL:NO /subsystem:console  LIB\c10.lib  LIB\gmock.lib  LIB\gtest.lib  LIB\gtest_main.lib  LIB\gtest.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ." 
Microsoft (R) Incremental Linker Version 14.23.28107.0 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3962448) pytorch_libtorch_linux_xenial_cuda9_cudnn7_py3_build (6/6)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74977734))
<details>
<summary>
<code>Dec 16 10:21:29 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected string-literal before ')' token</code>
</summary>

```
Dec 16 10:21:29                                                                  ^ 
Dec 16 10:21:29 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:105:65: error: expected string-literal before ')' token 
Dec 16 10:21:29 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp: In member function 'virtual void {anonymous}::test_if_constexpr::if_constexpr_returnTypeCanDiffer_withoutIdentityArg_Test::TestBody()': 
Dec 16 10:21:29 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:113:71: error: expected ',' before ')' token 
Dec 16 10:21:29      static_assert(std::is_same<std::string, decltype(a_string)>::value); 
Dec 16 10:21:29                                                                        ^ 
Dec 16 10:21:29 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:113:71: error: expected string-literal before ')' token 
Dec 16 10:21:29 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected ',' before ')' token 
Dec 16 10:21:29      static_assert(std::is_same<int64_t, decltype(an_int)>::value); 
Dec 16 10:21:29                                                                  ^ 
Dec 16 10:21:29 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected string-literal before ')' token 
Dec 16 10:21:29 make[2]: *** [c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o] Error 1 
Dec 16 10:21:29 make[1]: *** [c10/test/CMakeFiles/c10_C++17_test.dir/all] Error 2 
Dec 16 10:21:29 make[1]: *** Waiting for unfinished jobs.... 
Dec 16 10:21:29 c10/test/CMakeFiles/c10_C++17_test.dir/build.make:62: recipe for target 'c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o' failed 
Dec 16 10:21:29 make[2]: Leaving directory '/var/lib/jenkins/cpp-build/caffe2/build' 
Dec 16 10:21:29 CMakeFiles/Makefile2:2497: recipe for target 'c10/test/CMakeFiles/c10_C++17_test.dir/all' failed 
Dec 16 10:21:30 [ 24%] Linking CXX executable ../../bin/c10_DeviceGuard_test 
Dec 16 10:21:30 cd /var/lib/jenkins/cpp-build/caffe2/build/c10/test && /usr/bin/cmake -E cmake_link_script CMakeFiles/c10_DeviceGuard_test.dir/link.txt --verbose=1 
stics-color=always -Werror -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -g -fno-omit-frame-pointer -O0   -rdynamic  -Wl,-rpath  -Wl,/usr/lib/openmpi/lib  -Wl,--enable-new-dtags CMakeFiles/c10_DeviceGuard_test.dir/core/DeviceGuard_test.cpp.o  -o ../../bin/c10_DeviceGuard_test ../../lib/libc10.so ../../lib/libgmockd.a ../../lib/libgtestd.a ../../lib/libgtest_maind.a ../../lib/libgtestd.a -lpthread -Wl,-rpath,/var/lib/jenkins/cpp-build/caffe2/build/lib:  
Dec 16 10:21:30 make[2]: Leaving directory '/var/lib/jenkins/cpp-build/caffe2/build' 
```

</details>

### 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [caffe2_onnx_py3_6_clang7_ubuntu16_04_test](https://dr.pytorch.org/build-details.html?build_id=75047264)



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

smessmer committed:
Update on "Simplify code with decltype(auto)"


New c++14 feature we can use now

Differential Revision: [D18869644](https://our.internmc.facebook.com/intern/diff/D18869644/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify code with decltype(auto)"


New c++14 feature we can use now

Differential Revision: [D18869644](https://our.internmc.facebook.com/intern/diff/D18869644/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify code with decltype(auto)"


New c++14 feature we can use now

Differential Revision: [D18869644](https://our.internmc.facebook.com/intern/diff/D18869644/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify code with decltype(auto)"


New c++14 feature we can use now

Differential Revision: [D18869644](https://our.internmc.facebook.com/intern/diff/D18869644/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify code with decltype(auto)"


New c++14 feature we can use now

Differential Revision: [D18869644](https://our.internmc.facebook.com/intern/diff/D18869644/)

[ghstack-poisoned]

smessmer committed:
Update on "Simplify code with decltype(auto)"


New c++14 feature we can use now

Differential Revision: [D18869644](https://our.internmc.facebook.com/intern/diff/D18869644/)

[ghstack-poisoned]
